File: ts/src/relay/routes/participants.ts
import { Hono } from 'hono';
import type { RelayState, Participant } from '../app';

export function createParticipantsRoutes(state: RelayState) {
  const app = new Hono();

  // POST /rad/participants - 参加登録
  app.post('/rad/participants', async (c) => {
    const body = await c.req.json();

    // publicKey 検証
    if (!body.publicKey) {
      return c.json({ error: 'publicKey is required' }, 400);
    }

    // participantId: リクエストから取得、なければ displayName、それもなければ生成
    const participantId = body.participantId || body.displayName || `participant-${Date.now()}`;

    // 最初の参加者は Founder（またはリクエストで明示的に指定）
    const isFounder = body.isFounder !== undefined ? body.isFounder : state.participants.size === 0;

    const participant: Participant = {
      participantId,
      publicKey: body.publicKey,
      displayName: body.displayName,
      isFounder,
      joinedAt: Date.now(),
    };

    state.participants.set(participantId, participant);

    // Founder として登録
    if (isFounder) {
      state.founderTree.registerFromWrite('.', participantId);
    }

    return c.json(participant, 201);
  });

  // GET /rad/participants - 参加者一覧
  app.get('/rad/participants', (c) => {
    const participantList = Array.from(state.participants.values());
    return c.json(participantList, 200);
  });

  return app;
}

File: rust/src/relay/participants.rs
use axum::{Json, extract::State, http::StatusCode};
use serde::{Deserialize, Serialize};
use super::state::{SharedState, Participant};

#[derive(Debug, Deserialize)]
pub struct JoinRequest {
    #[serde(rename = "publicKey")]
    pub public_key: Option<String>,
    #[serde(rename = "displayName")]
    pub display_name: Option<String>,
}

// POST /rad/participants
pub async fn join(
    State(state): State<SharedState>,
    Json(body): Json<JoinRequest>,
) -> Result<(StatusCode, Json<Participant>), (StatusCode, Json<ErrorResponse>)> {
    // publicKey 検証
    let public_key = match body.public_key {
        Some(pk) => pk,
        None => {
            return Err((
                StatusCode::BAD_REQUEST,
                Json(ErrorResponse {
                    error: "publicKey is required".to_string(),
                }),
            ));
        }
    };

    // participantId 生成
    let participant_id = body.display_name.clone().unwrap_or_else(|| format!("participant-{}", chrono::Utc::now().timestamp_millis()));

    // 最初の参加者は Founder
    let mut participants = state.participants.lock().unwrap();
    let is_founder = participants.is_empty();

    let participant = Participant {
        participant_id: participant_id.clone(),
        public_key,
        display_name: body.display_name,
        is_founder,
        joined_at: chrono::Utc::now().timestamp_millis() as u64,
    };

    participants.push(participant.clone());

    // Founder として登録
    if is_founder {
        let mut founder_tree = state.founder_tree.lock().unwrap();
        founder_tree.register_from_write(".", &participant_id);
    }

    Ok((StatusCode::CREATED, Json(participant)))
}

// GET /rad/participants
pub async fn list(
    State(state): State<SharedState>,
) -> Json<Vec<Participant>> {
    let participants = state.participants.lock().unwrap();
    Json(participants.clone())
}

#[derive(Debug, Serialize)]
pub struct ErrorResponse {
    pub error: String,
}
